src/main/java/com/godaddy/sonar/ruby/simplecovrcov/SimpleCovRcovJsonParserImpl.java

```diff
@@ -11,49 +16,121 @@
+            reporter.addItem(buildReporterItem(reportItemMapEntry));
+        }
+        return reporter;
+    }
+
+    private ReporterItem buildReporterItem(Map.Entry reporterItemMapEntry) {
+        String filename = reporterItemMapEntry.getKey().toString();
+        JsonArray marksJsonArr = ((JsonArray)reporterItemMapEntry.getValue());
+        Collection<Mark> marks = new ArrayList<>();
+        for(JsonElement marksEl : marksJsonArr) {
+            marks.add(new Mark(marksEl.toString(), marksEl.isJsonNull()));
+        }
+        return new ReporterItem(filename, marks);
+    }
+
+    private Map<String, CoverageMeasuresBuilder> processReporter(Reporter reporter, Map<String, CoverageMeasuresBuilder> coveredFiles) {
+        for (ReporterItem reporterItem : reporter.getItems()) {
+            String filename = reporterItem.getFilename();
+            CoverageMeasuresBuilder fileCoverage = coveredFiles.getOrDefault(filename, CoverageMeasuresBuilder.create());
+            processReporterItem(reporterItem, fileCoverage);
+            coveredFiles.put(reporterItem.getFilename(), fileCoverage);
+        }
+        return coveredFiles;
+    }
+
+    private void processReporterItem(ReporterItem reporterItem, CoverageMeasuresBuilder fileCoverage) {
+        ArrayList<Mark> reporterItemMarks = (ArrayList<Mark>) reporterItem.getMarks();
+        for (int markId = 0; markId < reporterItemMarks.size(); markId++) {
+            Mark mark = reporterItemMarks.get(markId);
+            processMark(mark, markId, fileCoverage);
+        }
+    }
+
+    private void processMark(Mark mark, Integer index, CoverageMeasuresBuilder fileCoverage) {
+        if (mark.getIsNull()) { return; }
+        int hitsCount = mark.getAsLong().intValue();
+        int lineNumber = index + 1;
+        mergeHitsCounters(hitsCount, lineNumber, fileCoverage);
+    }
+
+    /**
+     * Merges hits count of a specified line with already existing coverage results for a given file.
+     *
+     * Coverage information for one file could be presented in several suites. So we have to merge all results per file.
+     * Considering how {@link org.sonar.api.measures.CoverageMeasuresBuilder#setHits} works this a little bit
+     * tricky method is needed: if you already have added some coverage info for the particular line, then you can not
+     * add more information. Please, visit method source code for more understanding. It's very small :-)
+     *
+     * @param hitsCount
+     * @param lineNumber
+     * @param fileCoverage
+     */
+    private void mergeHitsCounters(int hitsCount, int lineNumber, CoverageMeasuresBuilder fileCoverage) {
+        SortedMap<Integer, Integer> hitsByLine = Maps.newTreeMap();
+        hitsByLine.putAll(fileCoverage.getHitsByLine());
+        int oldHits = hitsByLine.getOrDefault(lineNumber, 0);
+        hitsByLine.put(lineNumber, oldHits + hitsCount);
+        fileCoverage.reset();
```

Comment:
Now I understand, then this should definitely be happening at the end of the process, you should create a Maps.newTreeMap containing this information and do the iteration once at the end. Otherwise this is an exponential runtime which will incur great time costs for bigger project. (Because we are iterating the entire report each time)
But I like the workaround :+1: